scene/gui/text_edit_menu.cpp

```diff
@@ -0,0 +1,214 @@
+	idx = m_menu->get_item_index(m_id);                                                                                  \
+	if (idx >= 0) {                                                                                                      \
+		m_menu->set_item_accelerator(idx, p_shortcut_keys_enabled ? _get_menu_action_accelerator(m_action) : Key::NONE); \
+		m_menu->set_item_disabled(idx, m_disabled);                                                                      \
+	}
+
+#define MENU_ITEM_ACTION(m_menu, m_id, m_action)                                                                         \
+	idx = m_menu->get_item_index(m_id);                                                                                  \
+	if (idx >= 0) {                                                                                                      \
+		m_menu->set_item_accelerator(idx, p_shortcut_keys_enabled ? _get_menu_action_accelerator(m_action) : Key::NONE); \
+	}
+
+#define MENU_ITEM_DISABLED(m_menu, m_id, m_disabled) \
+	idx = m_menu->get_item_index(m_id);              \
+	if (idx >= 0) {                                  \
+		m_menu->set_item_disabled(idx, m_disabled);  \
+	}
+
+#define MENU_ITEM_CHECKED(m_menu, m_id, m_checked) \
+	idx = m_menu->get_item_index(m_id);            \
+	if (idx >= 0) {                                \
+		m_menu->set_item_checked(idx, m_checked);  \
+	}
+
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_CUT, "ui_cut", !p_editable)
+	MENU_ITEM_ACTION(menu, MENU_COPY, "ui_copy")
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_PASTE, "ui_paste", !p_editable)
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_SELECT_ALL, "ui_text_select_all", !p_selecting_enabled)
+	MENU_ITEM_DISABLED(menu, MENU_CLEAR, !p_editable)
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_UNDO, "ui_undo", !p_editable || !p_has_undo)
+	MENU_ITEM_ACTION_DISABLED(menu, MENU_REDO, "ui_redo", !p_editable || !p_has_redo)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_INHERITED, p_text_direction == Control::TEXT_DIRECTION_INHERITED)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_AUTO, p_text_direction == Control::TEXT_DIRECTION_AUTO)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_LTR, p_text_direction == Control::TEXT_DIRECTION_LTR)
+	MENU_ITEM_CHECKED(menu_dir, MENU_DIR_RTL, p_text_direction == Control::TEXT_DIRECTION_RTL)
+	MENU_ITEM_CHECKED(menu, MENU_DISPLAY_UCC, p_draw_control_chars)
+	MENU_ITEM_DISABLED(menu, MENU_SUBMENU_INSERT_UCC, !p_editable)
+
+#undef MENU_ITEM_ACTION_DISABLED
+#undef MENU_ITEM_ACTION
+#undef MENU_ITEM_DISABLED
+#undef MENU_ITEM_CHECKED
+}
+
+void TextEditMenu::popup_at(const Point2 &p_point, bool p_grab_focus) {
+	if (menu) {
+		menu->set_position(p_point);
+		menu->reset_size();
+		menu->popup();
+		if (p_grab_focus) {
+			menu->grab_focus();
+		}
+	}
+}
+
+bool TextEditMenu::has_focus() {
+	return menu && menu->has_focus();
+}
+
+void TextEditMenu::set_text_direction(Control::TextDirection p_text_direction) {
+	if (menu_dir) {
```

Comment:
```suggestion
	if (!menu_dir) {
		return;
	}
```
Prefer early returns instead of nesting, when possible.

Reply:
`ERR_FAIL_NULL();`  added to: `set_text_direction`, `popup_at`, `set_draw_control_chars`, `add_option_selected_callback` and `add_focus_changed_callback`. Was already in `update`
Will be in next push.

Reply:
Turns out adding it to `set_text_direction` gave errors because of the CodeEditor. I edited `TextEdit` so that it now generates the menu if it hasn't already before calling that function.